File: src/jupyter_matlab_labextension/src/plugins/actions/actionFactory.ts
// Copyright 2025 The MathWorks, Inc.

import { BaseAction } from './baseAction';
import { ActionTypes } from './actionTypes';

import { ConvertAction } from './convertAction';
import { EditAction } from './editAction';
import { MatlabStatusAction } from './matlabStatusAction';
import { StartMatlabProxyAction } from './startMatlabProxyAction';
import { CheckFileExistsAction } from './checkFileExistsAction';
import { NudgeAction } from './nudgeAction';
import { NotebookPanel } from '@jupyterlab/notebook';

export class ActionFactory {
    static createAction (actionType: string, blocking: boolean, panel: NotebookPanel): BaseAction {
        switch (actionType) {
            case ActionTypes.CONVERT:
                return new ConvertAction(blocking);

            case ActionTypes.EDIT:
                return new EditAction(blocking);

            case ActionTypes.MATLAB_STATUS:
                return new MatlabStatusAction(blocking);

            case ActionTypes.START_MATLAB_PROXY:
                return new StartMatlabProxyAction(blocking);

            case ActionTypes.CHECK_FILE_EXISTS:
                return new CheckFileExistsAction(blocking);

            case ActionTypes.NUDGE:
                return new NudgeAction(blocking, panel);

            default:
                throw new Error(`Unknown action type: ${actionType}`);
        }
    }
}

File: src/jupyter_matlab_kernel/kernels/labextension_comm/actions/action_factory.py
# Copyright 2025 The MathWorks, Inc.

from typing import Union
from jupyter_matlab_kernel.kernels.labextension_comm.actions import (
    ConvertAction,
    EditAction,
    MatlabStatusAction,
    StartMatlabProxyAction,
    CheckFileExistsAction,
    UnknownAction,
)
from jupyter_matlab_kernel.kernels.labextension_comm.actions.types import ActionTypes


class ActionFactory:
    """
    ActionFactory class for determining and returning the appropriate Action class.

    This class provides a static method to decide between different Action
    implementations based on the supplied Action Type.
    """

    @staticmethod
    def create_action(action_type, kernel) -> Union[
        ConvertAction,
        EditAction,
        MatlabStatusAction,
        StartMatlabProxyAction,
        UnknownAction,
    ]:
        """Determines and returns the appropriate Action to use

        Args:
            action_type (ActionTypes): Type of the action
            kernel (BaseMATLABKernel): The MATLAB Kernel being used

        Returns:
            Union[ConvertAction, EditAction, MatlabStatusAction, StartMatlabProxyAction, UnknownAction]: The appropriate Action class based on the provided action_type.
        """
        if ActionTypes.CONVERT.value == action_type:
            return ConvertAction(kernel)

        elif ActionTypes.EDIT.value == action_type:
            return EditAction(kernel)

        elif ActionTypes.MATLAB_STATUS.value == action_type:
            return MatlabStatusAction(kernel)

        elif ActionTypes.START_MATLAB_PROXY.value == action_type:
            return StartMatlabProxyAction(kernel)

        # Add more elif conditions as and when new action types are introduced
        elif ActionTypes.CHECK_FILE_EXISTS.value == action_type:
            return CheckFileExistsAction(kernel)

        else:
            return UnknownAction(kernel)
